ktor-features/ktor-auth-jwt/jvm/src/io/ktor/auth/jwt/JWTPrincipal.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.auth.jwt
+
+import com.auth0.jwt.exceptions.*
+import com.auth0.jwt.interfaces.*
+import io.ktor.auth.*
+import java.util.*
+import kotlin.reflect.*
+
+interface JWToken: Credential, Principal {
+    /**
+     * Get the value of the "iss" claim, or null if it's not available.
+     */
+    val issuer: String?
+
+    /**
+     * Get the value of the "sub" claim, or null if it's not available.
+     */
+    val subject: String?
+
+    /**
+     * Get the value of the "aud" claim, or null if it's not available.
+     */
+    val audience: List<String>?
+
+    /**
+     * Get the value of the "exp" claim, or null if it's not available.
+     */
+    val expiresAt: Date?
+
+    /**
+     * Get the value of the "nbf" claim, or null if it's not available.
+     */
+    val notBefore: Date?
+
+    /**
+     * Get the value of the "iat" claim, or null if it's not available.
+     */
+    val issuedAt: Date?
+
+    /**
+     * Get the value of the "jti" claim, or null if it's not available.
+     */
+    val jwtId: String?
+
+    /**
+     * Attempt to locate a String claim with the specified name, null if it's not available or not a String
+     */
+    operator fun get(name: String): String?
+
+    /**
+     * Attempt to locate a claim with the specified name, and deserialize it as [clazz]
+     */
+    fun <T: Any> getClaim(name: String, clazz: KClass<T>): T?
+
+    /**
+     * Attempt to locate a claim with the specified name, and deserialize it as List<[clazz]>
+     */
+    fun <T: Any> getListClaim(name: String, clazz: KClass<T>): List<T>?
+}
+
+inline fun <reified T: Any> JWToken.getClaim(name: String): T? {
+    return getClaim(name, T::class)
+}
+
+inline fun <reified T: Any> JWToken.getListClaim(name: String): List<T>? {
+    return getListClaim(name, T::class)
+}
+
+/**
+ * JWT payload from auth0-jwt
+ */
+open class Auth0Jwt(private val payload: Payload) : JWToken {
```

Comment:
What is the reason to have it public and open? I would have it internal instead. What usecase for inheriting it?

Reply:
I suppose it's not strictly necessary in a kotlin world, but one could do so in order to cache non-standard claims from the token as fields instead of querying them every time